src/coreclr/jit/importer.cpp

```diff
@@ -3647,6 +3647,110 @@ const char* Compiler::impGetIntrinsicName(CorInfoIntrinsics intrinsicID)
+    {
+        return nullptr;
+    }
+
+    CORINFO_CLASS_HANDLE fieldOwnerHnd = info.compCompHnd->getFieldClass(fieldToken);
+
+    CORINFO_CLASS_HANDLE fieldClsHnd;
+    var_types fieldElementType = JITtype2varType(info.compCompHnd->getFieldType(fieldToken, &fieldClsHnd, fieldOwnerHnd));
+    unsigned totalFieldSize;
+
+    // Most static initialization data fields are of some structure, but it is possible for them to be of various primitive types as well
+    if (fieldElementType == var_types::TYP_STRUCT)
+    {
+        totalFieldSize = info.compCompHnd->getClassSize(fieldClsHnd);
+    }
+    else
+    {
+        totalFieldSize = genTypeSize(fieldElementType);
+    }
+
+    // Limit to primitive or enum type - see ArrayNative::GetSpanDataFrom()
+    CORINFO_CLASS_HANDLE targetElemHnd  = sig->sigInst.methInst[0];
+    if (info.compCompHnd->getTypeForPrimitiveValueClass(targetElemHnd) == CORINFO_TYPE_UNDEF)
+    {
+        return nullptr;
+    }
+
+    const unsigned targetElemSize = info.compCompHnd->getClassSize(targetElemHnd);
+    assert(targetElemSize != 0);
+
+    const unsigned count = totalFieldSize / targetElemSize;
+    if (count == 0)
+    {
+        return nullptr;
+    }
+
+    void* data = info.compCompHnd->getArrayInitializationData(fieldToken, totalFieldSize);
+    if (!data)
+    {
+        return nullptr;
+    }
+
+    //
+    // Ready to commit to the work
+    //
+
+    impPopStack();
+
+    // Turn count and pointer value into constants.
+    GenTree* spanElemCount = gtNewIconNode(count, TYP_INT);
+    GenTree* spanPointerField = gtNewIconHandleNode((size_t)data, GTF_ICON_CONST_PTR);
+
+    // Construct ReadOnlySpan<T> to return.
+    unsigned spanTTemp = lvaGrabTemp(true DEBUGARG("ReadOnlySpan<T> for CreateSpan<T>"));
+    lvaSetStruct(spanTTemp, sig->retTypeClass, false);
+    GenTree* addrOfSpanTTempPointer = gtNewOperNode(GT_ADDR, TYP_BYREF, impCreateLocalNode(spanTTemp DEBUGARG(0))); // HACKATHON OFFSET not set usefully here
+    GenTree* dereffedSpanTTempPointer = gtNewOperNode(GT_IND, TYP_BYREF, addrOfSpanTTempPointer);
+    GenTree* gtAssignedSpanTTempPointer = gtNewAssignNode(dereffedSpanTTempPointer, spanPointerField);
+
+    GenTree* addrOfSpanTTempPointer2 = gtNewOperNode(GT_ADDR, TYP_BYREF, impCreateLocalNode(spanTTemp DEBUGARG(0))); // HACKATHON OFFSET not set usefully here
+    GenTree* addrOfSpanTTempSize = gtNewOperNode(GT_ADD, TYP_BYREF, addrOfSpanTTempPointer2, gtNewIconNode(TARGET_POINTER_SIZE, TYP_I_IMPL)); // HACKATHON OFFSET not set usefully here
+    GenTree* dereffedSpanTTempSize = gtNewOperNode(GT_IND, TYP_INT, addrOfSpanTTempSize);
+    GenTree* gtAssignedSpanTTempSize = gtNewAssignNode(dereffedSpanTTempSize, spanElemCount);
+
+    GenTree* gtCommaAssignPointerAndGetResult = gtNewOperNode(GT_COMMA, TYP_STRUCT, gtAssignedSpanTTempPointer, impCreateLocalNode(spanTTemp DEBUGARG(0)));
+    return gtNewOperNode(GT_COMMA, TYP_STRUCT, gtAssignedSpanTTempSize, gtCommaAssignPointerAndGetResult);
```

Comment:
The most important thing CQ-wise in this code is to make it easy to promote the span. I see this code does indirect stores to the fields - I think you need actual field stores, either `ASG(LCL_FLD span, count/pointer)`, or the more elaborate `ASG(FIELD(ADDR(span)), count/pointer)` (I believe the field handles can be acquired via `getFieldInClass`). It would be nice to not have to waste IR on the latter form, but it needs checking against what `LocalAddressVisitor` recognizes today - it may miss `LCL_FLD`s as they are rarely created this early.

It would be nice to avoid `COMMA`s too (they take up more space than statements), but that is probably not worth the complexity.

I note as well that it seems inconsequential what offsets you supply for the locals - they are `DEBUG`-only after all. `BAD_IL_OFFSET`/`opcodeOffs` may just do.

Reply:
So I have checked the codegen with this version:
```c++
// Turn count and pointer value into constants.
GenTree* lengthValue  = gtNewIconNode(count, TYP_INT);
GenTree* pointerValue = gtNewIconHandleNode((size_t)data, GTF_ICON_CONST_PTR);

// Construct ReadOnlySpan<T> to return.
CORINFO_CLASS_HANDLE spanHnd     = sig->retTypeClass;
unsigned             spanTempNum = lvaGrabTemp(true DEBUGARG("ReadOnlySpan<T> for CreateSpan<T>"));
lvaSetStruct(spanTempNum, spanHnd, false);

CORINFO_FIELD_HANDLE pointerFieldHnd = info.compCompHnd->getFieldInClass(spanHnd, 0);
CORINFO_FIELD_HANDLE lengthFieldHnd  = info.compCompHnd->getFieldInClass(spanHnd, 1);

GenTreeLclFld* pointerField = gtNewLclFldNode(spanTempNum, TYP_BYREF, 0);
pointerField->SetFieldSeq(GetFieldSeqStore()->CreateSingleton(pointerFieldHnd));
GenTree* pointerFieldAsg = gtNewAssignNode(pointerField, pointerValue);

GenTreeLclFld* lengthField = gtNewLclFldNode(spanTempNum, TYP_INT, TARGET_POINTER_SIZE);
lengthField->SetFieldSeq(GetFieldSeqStore()->CreateSingleton(lengthFieldHnd));
GenTree* lengthFieldAsg = gtNewAssignNode(lengthField, lengthValue);

// Now chain them together via COMMAs.
GenTree* result = impCreateLocalNode(spanTempNum DEBUGARG(0));
result          = gtNewOperNode(GT_COMMA, TYP_STRUCT, pointerFieldAsg, result);
result          = gtNewOperNode(GT_COMMA, TYP_STRUCT, lengthFieldAsg, result);

return result;
```
And it does work as intended CQ-wise. Egor suggests we intrinsify the `out`-style version of the method, that seems advantageous to me as well, it should result in a bit smaller IR than this version. The only problem I see is if we leave `CreateSpan` non-intrinsified and leave it to the inlining to clean things up, it'll be worse TP-wise (and presumably we're not making `GetSpanData` public for Roslyn to call... we could re-review the API of course, but that seems not worth it for this Jit micro-optimization).

Reply:
> Egor suggests we intrinsify the out-style version of the method, that seems advantageous to me as well, it should result in a bit smaller IR than this version

The problem with intrinsifying the internal out-style helper is that you will have hard time picking up the field token from the IL stream. I think you would have to intrinsify the public entrypoint to do that as well, and then blend the two together after the fact somehow.

I think it is easier and better to just intrinsify the public entrypoint.

Reply:
> So I have checked the codegen with this version:

Nice! Looks good to me, I'd only use 
```
impAppendTree(lengthFieldAsg, (unsigned)CHECK_SPILL_NONE, impCurStmtOffs);
impAppendTree(pointerFieldAsg, (unsigned)CHECK_SPILL_NONE, impCurStmtOffs);

return impCreateLocalNode(spanTempNum DEBUGARG(0));
```
instead of:
```
GenTree* result = impCreateLocalNode(spanTempNum DEBUGARG(0));
result          = gtNewOperNode(GT_COMMA, TYP_STRUCT, pointerFieldAsg, result);
result          = gtNewOperNode(GT_COMMA, TYP_STRUCT, lengthFieldAsg, result);

return result;
```

Reply:
This fix is included in the PR